File: Java/프로그래머스/1/72410. 신규 아이디 추천/신규 아이디 추천.java
class Solution {
    public String solution(String new_id) {
        String answer = new_id;
        // step 1
        answer = answer.toLowerCase();
            
        // step 2
        StringBuilder sb = new StringBuilder();
        for (char c : answer.toCharArray()) {
            if ((c >= '0' && c <= '9') || (c >= 'a' && c <= 'z') || c == '-' || c == '_' || c == '.') {
                sb.append(String.valueOf(c));
            }
        }
        answer = sb.toString();
        
        // step 3
        while (answer.contains("..")) {
            answer = answer.replace("..", ".");
        }
        
        //step 4
        if (answer.length() > 0 && answer.charAt(0) == '.') {
            answer = answer.substring(1);
        }
        if (answer.length() > 0 && answer.charAt(answer.length() - 1) == '.') {
            answer = answer.substring(0, answer.length() - 1);
        }
        
        
        //step 5
        if (answer.isEmpty()) {
            answer = "a";
        }
        
        // step 6
        if (answer.length() >= 16) {
            answer = answer.substring(0, 15);
            if (answer.charAt(answer.length() - 1) == '.') {
                answer = answer.substring(0, 14);
            }
        }
        
        if (answer.length() <= 2) {
            sb = new StringBuilder(answer);
            char lastChar = answer.charAt(answer.length()-1);
            while(sb.length() < 3) {
                sb.append(String.valueOf(lastChar));
            } 
            answer = sb.toString();
        }
        
        
        return answer;
    }
}

File: Swift/프로그래머스/1/72410. 신규 아이디 추천/신규 아이디 추천.swift
import Foundation

func solution(_ new_id:String) -> String {
    var answer = new_id
    
    // step 1
    answer = answer.lowercased()
    
    // step 2
    let allowedChars = "abcdefghijklmnopqrstuvwxyz0123456789-_."
    answer = answer.filter { allowedChars.contains($0) }.map { String($0) }.joined()
    
    // step 3
    while answer.contains("..") {
        answer = answer.replacingOccurrences(of: "..", with: ".")
    }
    
    // step 4
    if answer.hasPrefix(".") {
        answer.removeFirst()
    }
    if answer.hasSuffix(".") {
        answer.removeLast()
    }
    
    // step 5
    if answer.isEmpty {
        answer = "a"
    }
    
    // step 6
    if answer.count >= 16 {
        let index = answer.index(answer.startIndex, offsetBy: 15)
        answer = String(answer[..<index])
        
        if answer.hasSuffix(".") {
            answer.removeLast()
        }
    }
    
    // step 7: 길이가 2자 이하라면, 마지막 문자를 길이가 3이 될 때까지 반복해서 붙임
    while answer.count <= 2 {
        answer.append(answer.last!)
    }
    
    return answer
}